crates/circuit/src/parameter_table.rs

```diff
@@ -10,164 +10,395 @@
+            let info = &self.by_uuid[uuid];
+            if let Some(vec) = info.element.as_ref() {
+                (&self.vectors[&vec.vector_uuid].name, vec.index)
+            } else {
+                (&info.name, 0)
+            }
+        })
+    }
+
+    /// Add a use of a parameter to the table.
+    pub fn add_use(
+        &mut self,
+        uuid: ParameterUuid,
+        usage: ParameterUse,
+    ) -> Result<(), ParameterTableError> {
+        self.by_uuid
+            .get_mut(&uuid)
+            .ok_or(ParameterTableError::ParameterNotTracked(uuid))?
+            .uses
+            .insert(usage);
         Ok(())
     }
 
-    pub fn discard_references(
+    /// Return a use of a parameter.
+    ///
+    /// If the last use a parameter is discarded, the parameter is untracked.
+    pub fn remove_use(
         &mut self,
-        uuid: u128,
-        inst_index: usize,
-        param_index: usize,
-        name: String,
-    ) {
-        if let Some(refs) = self.table.get_mut(&uuid) {
-            if refs.__len__() == 1 {
-                self.table.remove(&uuid);
-                self.names.remove(&name);
-                self.uuid_map.remove(&uuid);
-            } else {
-                refs.discard(inst_index, param_index);
+        uuid: ParameterUuid,
+        usage: ParameterUse,
+    ) -> Result<(), ParameterTableError> {
+        let Entry::Occupied(mut entry) = self.by_uuid.entry(uuid) else {
+            return Err(ParameterTableError::ParameterNotTracked(uuid));
+        };
+        let info = entry.get_mut();
+        if !info.uses.remove(&usage) {
+            return Err(ParameterTableError::UsageNotTracked(usage));
+        }
+        if info.uses.is_empty() {
+            self.by_name.remove(&info.name);
+            if let Some(vec) = info.element.as_ref() {
+                let Entry::Occupied(mut vec_entry) = self.vectors.entry(vec.vector_uuid) else {
+                    unreachable!()
+                };
+                vec_entry.get_mut().refcount -= 1;
+                if vec_entry.get().refcount == 0 {
+                    vec_entry.remove_entry();
+                }
             }
+            self.order.clear();
+            self.py_parameters = None;
+            entry.remove_entry();
         }
+        Ok(())
     }
-}
 
-#[pymethods]
-impl ParamTable {
-    #[new]
-    pub fn new() -> Self {
-        ParamTable {
-            table: HashMap::new(),
-            names: HashMap::new(),
-            uuid_map: HashMap::new(),
+    /// Remove a parameter from the table, returning the tracked uses of it.
+    pub fn pop(
+        &mut self,
+        uuid: ParameterUuid,
+    ) -> Result<HashSet<ParameterUse>, ParameterTableError> {
+        let info = self
+            .by_uuid
+            .remove(&uuid)
+            .ok_or(ParameterTableError::ParameterNotTracked(uuid))?;
+        self.by_name
+            .remove(&info.name)
+            .expect("each parameter should be tracked by both UUID and name");
```

Comment:
This isn't a user-facing warning; it's triggering an internal panic if one of the core invariants of the `ParameterTable` is violated.  It _shouldn't_ be possible for it to occur at all, which is why we're using `Option::expect`, rather than mapping it into a Python-space error.

If this triggers, it'll be because a developer (maybe me) made a mistake, and the backtrace will point immediately back to this line of code.